timemachine/cpp/src/kernels/k_barostat.cuh

```diff
@@ -121,20 +123,23 @@ void __global__ k_decide_move(
     const int num_molecules,
     const double kt,
     const double pressure,
-    const RealType *__restrict__ rand, // [2] Use second value
-    RealType *__restrict__ d_volume_delta,
-    double *__restrict__ d_volume_scale,
-    const __int128 *__restrict__ d_init_u,
-    const __int128 *__restrict__ d_final_u,
-    double *__restrict__ d_box,
-    const double *__restrict__ d_box_output,
-    double *__restrict__ d_x,
-    const double *__restrict__ d_x_output,
-    int *__restrict__ num_accepted,
-    int *__restrict__ num_attempted) {
+    const RealType *__restrict__ rand,           // [2] Use second value
+    const RealType *__restrict__ d_volume,       // [1]
+    const RealType *__restrict__ d_volume_delta, // [1]
+    double *__restrict__ d_volume_scale,         // [1]
+    const __int128 *__restrict__ d_init_u,       // [1]
+    const __int128 *__restrict__ d_final_u,      // [1]
+    double *__restrict__ d_box,                  // [3*3]
+    const double *__restrict__ d_box_output,     // [3*3]
+    double *__restrict__ d_x,                    // [N*3]
+    const double *__restrict__ d_x_output,       // [N*3]
+    int *__restrict__ num_accepted,              // [1]
+    int *__restrict__ num_attempted              // [1]
+) {
     int idx = blockIdx.x * blockDim.x + threadIdx.x;
 
-    const RealType volume = d_box[0 * 3 + 0] * d_box[1 * 3 + 1] * d_box[2 * 3 + 2];
+    // Don't compute volume from the box. It leads to a race condition since `d_box` is updated in this kernel
```

Comment:
I'd also be in favor of renaming `d_x_output` and `d_box_output` to  `d_x_proposed` and `d_box_proposed`.

This actually tripped me up on first read since usually things that are `_output` are *not* `const`-qualified, here, it's actually an `_input` to the function. I had to read it a second time to catch it.